### DISCUSSION TITLE
error SyntaxError: Unexpected end of JSON input

### DISCUSSION THREAD

**Discussion by JinMABrett** [nextauthjs/next-auth#7887] [Help] [answered]
## Project INfo
Next.js 13.4 using app router > Next auth 4.22.1 > Vercel deployment. 
NEXTAUTH_URL and NERXTAUTH_SECRET are set as env variables on vercel. 

## UPDATE
So the vercel log error i'm getting seems to be triggering on the 
```javascript
const body = await request.json();
```
Line in the "api/login" page. On localhost for whatever reason the request seems to be working just fine, but on live vercel server for whatever reason, it seems to not be sending the credentials and possibly is empty and throwing the error listed below. I have no idea how to check or fix this though as i'm a new programmer.  Any help would be appreciated. Thank you

## The Problem
When I have my project running on localhost, everything seems to work fine. However when I push to vercel, 100% of my login attempts bring me to an error page that just says "Error" with the url under it. 
    
## test website Link
Here is my test deploy. The login button is actually the colored (green) button in the middle of the hero section that says "sign in". (was the easiest place to place it for testing). 

https://jinmatestapi.vercel.app/

# Vercel log error
Was able to track down this error triggering on api/login on vercel logs.

```vbnet
- error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6571:19)
    at successSteps (node:internal/deps/undici/undici:6545:27)
    at node:internal/deps/undici/undici:1211:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```    

# The Code
##    **Here is my test api/[...nextauth] route**
```javascript
import NextAuth from "next-auth/next";
import CredentialsProvider from "next-auth/providers/credentials";

const handler = NextAuth({
  providers: [
    CredentialsProvider({
      name: "Credentials",
      credentials: {
        username: { label: "Username", type: "text", placeholder: "test" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials, req) {
        const res = await fetch(`${process.env.NEXTAUTH_URL}/api/login`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: credentials && credentials.username,
            password: credentials && credentials.password,
          }),
        });


        const user = await res.json();

        if (user) {
          return user;
        }
          return null;
      },
    }),
  ],
  callbacks: {
    async jwt({ token, user }) {
      return { ...token, ...user };
    },

    async session({ session, token }) {
      session.user = token;
      return session;
    },
  },
});

export { handler as GET, handler as POST };
```   

##    **Here is my test api/login route**
    
```javascript
// This is where we want to get the user from the database and compare the password

import { signJwtAccessToken } from 'jinma/auth/jwt';


export async function POST(request) {
  const body = await request.json();

  const user = {
    id: 1,
    username: 'test',
    name: 'test',
    password: 'test'
  }


if (user && (body.password === user.password) ) {
  const { password, ...userWithoutPass } = user;
  const accessToken = signJwtAccessToken(userWithoutPass);
  const result = {
    ...userWithoutPass,
    accessToken,
  };
  console.log("RESULT", result);
  return new Response(JSON.stringify(result));
} 
  return new Response(JSON.stringify(null));
}

```

## Also here is the JWT code for full context just in case. 
```javascript
import jwt from "jsonwebtoken";

const DEFAULT_SIGN_OPTION = {
  expiresIn: "30d",
};

function signJwtAccessToken(payload, options = DEFAULT_SIGN_OPTION) {
  const secret_key = 'testkeykjdlkfajsdlkfjlasdkfjasldkflkasdjfas';
  const token = jwt.sign(payload, secret_key, options);
  return token;
}

function verifyJwt(token) {
  try {
    const secret_key = 'testkeykjdlkfajsdlkfjlasdkfjasldkflkasdjfas';
    const decoded = jwt.verify(token, secret_key);
    return decoded;
  } catch (error) {
    console.log(error);
    return null;
  }
}

export { signJwtAccessToken, verifyJwt };

```

### How to reproduce ☕️

Setup next.js 13.4 project as shown and push to vercel then try and log in.

**Comment by jamesallchin-rb** [comment]
Same for me

**Comment by jamesallchin-rb** [comment, marked as answer]
OK. I faced exactly the same problem. Essentially trying to use `fetch` inside of custom next-auth `CredentialsProvider` - in order to validate the credentials with my own API route (a Next JS app router route from within the same Next JS app). 

Works fine locally. Does not work on Vercel. I confirmed that the body is empty when it reaches my own API route for validating the credentials.

Very likely to do with a difference in node version (and therefore node `fetch` version) between my localhost and vercel. 

It all points to `fetch` problems. Replacing `fetch` with `axios` resolves the issue.

Therefore replace your fetch call e.g.

```
        const res = await fetch(`${process.env.NEXTAUTH_URL}/api/login`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: credentials && credentials.username,
            password: credentials && credentials.password,
          }),
        });
```

With an axios call. 

```
        const res = await axios({
          method: 'POST',
          url: `${process.env.NEXTAUTH_URL}/api/login`,
          data: credentials,
          headers: {
            'Content-Type': 'application/json',
          },
        });
```

**Reply by starball5** [reply]
Related on Stack Overflow: [Next-Auth works on local but not on vercel - SyntaxError: Unexpected end of JSON input](https://stackoverflow.com/q/76568550/11107541)

**Comment by JinMABrett** [comment]
Thank you so much. This seems to have done the trick. I really appreciate it, i've been tearing my hair out for like a month over this.
